### PR TITLE
feat: relay v2 discovery (go-libp2p v0.19.0)

### PR DESCRIPTION
**Pull request by marten-seemann**
Release Notes: https://github.com/libp2p/go-libp2p/releases/tag/v0.19.0

go-libp2p v0.19.0 implements circuit v2 relay discovery:  AutoRelays now reads peers from a channel (provided by autorelay.WithPeerSource), tests if those peers speak the relay v2 protocol and tries to obtain a reservation with these nodes.

This PR gets the closest peers from the DHT, and provides them to AutoRelay on a regular basis.
 
By default, AutoRelay does not use Circuit Relay v1 nodes any more. Circuit v1 support is only enabled when static relays are in use.

----

## TODO

- [x] cleanup legacy config around relays
- [x] decide how often we feed new peers to autorelay discovery
- [x] include [`Peering.Peers`](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#peeringpeers) in autorelay discovery
- [x] clarify mplex situation in js-ipfs
- [x] fix  tests
  - [x] panic - https://github.com/ipfs/go-ipfs/pull/8868/commits/747c1140be7d5d1732cc0d59d306dea38fb9a131
  - [x] pathfinder – https://github.com/ipfs/go-ipfs/pull/8868/commits/2ec71d1cfe635b1b16cd749fad66dfce3701ddbd
  - [x] missing `/plaintext/2.0.0` when `--disable-transport-encryption`  is passed (failing socatt test in `t0061-daemon-opts.sh`)
  - [x] circuit v1 (flaky)
  - [x] IPNS regression in `t0160-resolve.sh` test 
    - [x] fix in https://github.com/ipfs/go-namesys/pull/32
    - [x]  switch to release version of go-namesys with  https://github.com/ipfs/go-namesys/pull/32
- [x] TBD: what to do with  hole punching?
  - descoped from this PR, moved to https://github.com/ipfs/go-ipfs/pull/8748 

### BREAKING CHANGES

See `CHANGELOG.md`

### CODE REVIEW

**Comment by lidel** [comment]
@marten-seemann anything changed that could make `/plaintext/2.0.0` could stop working? 

`test/sharness/t0061-daemon-opts.sh` has a test with `--disable-transport-encryption`  
where
```
/multistream/1.0.0
ls
```
is expected to produce 
```
/multistream/1.0.0
/plaintext/2.0.0
```
but with go-libp2p 0.19 it gets 

```
/multistream/1.0.0
na
```

File: test/sharness/t0182-circuit-relay.sh
Comment (lidel) [inline test/sharness/t0182-circuit-relay.sh:66]: Tests in  `go-ipfs/test/sharness/t0182-circuit-relay.sh`   are flaky.

`ipfsi 2 swarm connect /p2p/$PEERID_1/p2p-circuit/p2p/$PEERID_0`   sometimes fails with:
```
NO_RESERVATION (204)
Error: connect 12D3KooWLuY1Q13o6Q91NoJn4Qv7RSq5rYS6uC4YCzKCJ3S8GnwQ failure: failed to dial 12D3KooWLuY1Q13o6Q91NoJn4Qv7RSq5rYS6uC4YCzKCJ3S8GnwQ:
  * [/p2p/12D3KooWNfs8uFpQf6NnseZKLyS3c8EFppGtUGejHpViBeg8f7Vt/p2p-circuit] error opening relay circuit: NO_RESERVATION (204)`
```

- it happens randomly, bumping it to 10 seconds does not help much, it still fails sometimes.
- shutting down nodes and starting them again fixes the problem
  -  it feels like covering up some underlying racy bug:   once `NO_RESERVATION` error is returned, the relay will never work (until reboot of entire testbed).

**Review by lidel — COMMENTED** [review]
Rebased, CI is green, if we want we could merge this for 0.13-rc1 as-is.


Caveat: `test/sharness/t0182-circuit-relay.sh` became flaky – could be a bug? but we could tackle it after RC1 if needed  – see comment below.

File: core/node/libp2p/routing.go
Comment (lidel) [inline core/node/libp2p/routing.go:208]: :information_source:  This will feed peers to AutoRelay every 15s, and then exponentially back off until it happens once an hour. 
Should be okay solution until we have better mechanism in go-libp2p (where libp2p asks for peers, instead of being fed)

I used `github.com/cenkalti/backoff/v4` because it was already an indirect dependency.

File: core/node/libp2p/routing.go
Comment (lidel) [inline core/node/libp2p/routing.go:223]: :information_source: this allows people to reuse already trusted peers from `Peering.Peers` in addition to peers discovered via `dht.WAN.GetClosestPeers`

File: test/sharness/t0182-circuit-relay.sh
Comment (lidel) [inline test/sharness/t0182-circuit-relay.sh:66]: :information_source:  below is not a blocker,  imo this PR can be merged and if this is a real problem we will figure it out with  RC1 being used in the real world.

:warning:  Tests in  `go-ipfs/test/sharness/t0182-circuit-relay.sh`   became  flaky.

`ipfsi 2 swarm connect /p2p/$PEERID_1/p2p-circuit/p2p/$PEERID_0`   sometimes fails with:
```
NO_RESERVATION (204)
Error: connect 12D3KooWLuY1Q13o6Q91NoJn4Qv7RSq5rYS6uC4YCzKCJ3S8GnwQ failure: failed to dial 12D3KooWLuY1Q13o6Q91NoJn4Qv7RSq5rYS6uC4YCzKCJ3S8GnwQ:
  * [/p2p/12D3KooWNfs8uFpQf6NnseZKLyS3c8EFppGtUGejHpViBeg8f7Vt/p2p-circuit] error opening relay circuit: NO_RESERVATION (204)`
```

- it happens randomly, bumping it to 10 seconds does not help much, it still fails sometimes.
- shutting down nodes and starting them again fixes the problem
  - ...but it feels like covering up some underlying racy bug:   once `NO_RESERVATION` error is returned, the relay will never work, no matter how long we wait and retry again  (only way to fix it is to reboot of the relay node /  testbed). 



Repro (runs tests until they error):

```console
$ killall ipfs ; i=0; while ./t0182-circuit-relay.sh -v; do echo " -----> run no. $i <----"; sleep 1; ((i=i+1)) ; done
```

File: go.mod
Comment (lidel) [inline go.mod:55]: - [ ] TODO: needs https://github.com/ipfs/go-namesys/pull/31

**Review by lidel — APPROVED** [review]
I believe this is ready for final review/merging. 

Should be good enough for go-ipfs 0.13-rc1 – see key comments below.

File: test/sharness/t0182-circuit-relay.sh
Comment (lidel) [reply test/sharness/t0182-circuit-relay.sh:66]: cc @marten-seemann  for visibility – lmk if I should fill a bug in go-libp2p, or how I can provide more details

File: core/node/libp2p/routing.go
Comment (lidel) [reply core/node/libp2p/routing.go:237]: Imo not worth it, because this entire func will be redone after go-libp2p 0.20 ships.

File: core/node/libp2p/routing.go
Comment (lidel) [reply core/node/libp2p/routing.go:198]: I got paranoid because of panics around go-libp2p 0.19 (e.g. https://github.com/libp2p/go-libp2p/pull/1467), and since this is bolt-on temporary feeder (that may get refactored multiple times) felt prudent to decrease blast radius.

**Comment by lidel** [comment]
Rebased, waiting for CI.

File: core/node/libp2p/routing.go
Comment (guseggert) [reply core/node/libp2p/routing.go:200]: seems like overkill but okay :)

File: test/sharness/t0182-circuit-relay.sh
Comment (lidel) [reply test/sharness/t0182-circuit-relay.sh:66]: Continued in https://github.com/ipfs/go-ipfs/pull/8967

**Pull request merged**
